File: scripts/gen-android-icons.py
#!/usr/bin/env python3
"""Regenerate Android launcher icons (Pillow-based twin of gen-android-icons.js).

Prefer the JS script (zero dependencies, what CI and the docs use); this one
exists for anyone working in Python. Both produce the same artwork:

- Adaptive foreground: the BARE campfire mark on transparency, inside the
  72dp safe circle of the 108dp layer.
- Legacy + round icons: the badge (mark on the theme-colored circle) at full
  size, so old and modern launchers agree.
- Adaptive background color: the theme navy.

Requires Pillow:  pip install pillow
Run from the repo root:  python3 scripts/gen-android-icons.py
"""
import os
import sys

try:
    from PIL import Image
except ImportError:
    sys.exit("Pillow required: pip install pillow")

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MARK = os.path.join(ROOT, "public", "icons", "campfire-logo.png")
BADGE = os.path.join(ROOT, "public", "icons", "campfire-badge.png")
RES = os.path.join(ROOT, "app", "src-tauri", "gen", "android", "app", "src", "main", "res")
BG = (0x1A, 0x1D, 0x29, 255)  # app background / theme-color

DENS = {"mdpi": 1, "hdpi": 1.5, "xhdpi": 2, "xxhdpi": 3, "xxxhdpi": 4}


def load(path, trim=True):
    im = Image.open(path).convert("RGBA")
    if not trim:
        return im
    bbox = im.getbbox()
    return im.crop(bbox) if bbox else im


def place(canvas, art, frac):
    w = int(min(canvas.size) * frac)
    lg = art.resize((w, int(w * art.height / art.width)), Image.LANCZOS)
    canvas.alpha_composite(lg, ((canvas.width - lg.width) // 2, (canvas.height - lg.height) // 2))
    return canvas


def main():
    mark = load(MARK)
    badge = load(BADGE, trim=False)  # keep the badge's own transparent margin
    for name, scale in DENS.items():
        d = os.path.join(RES, f"mipmap-{name}")
        os.makedirs(d, exist_ok=True)
        # adaptive foreground layer: 108dp, bare mark inside the safe circle
        fg_px = int(108 * scale)
        fg = Image.new("RGBA", (fg_px, fg_px), (0, 0, 0, 0))
        place(fg, mark, 66 / 108)
        fg.save(os.path.join(d, "ic_launcher_foreground.png"))
        # legacy + round icons: the badge at full size (flat, resp. masked)
        px = int(48 * scale)
        for fname, full_bleed in (("ic_launcher.png", True), ("ic_launcher_round.png", False)):
            canvas = Image.new("RGBA", (px, px), BG if full_bleed else (0, 0, 0, 0))
            place(canvas, badge, 1.0)
            canvas.save(os.path.join(d, fname))
    with open(os.path.join(RES, "values", "ic_launcher_background.xml"), "w") as f:
        f.write('<?xml version="1.0" encoding="utf-8"?>\n<resources>\n  <color name="ic_launcher_background">#1a1d29</color>\n</resources>')
    print("android launcher icons regenerated")


if __name__ == "__main__":
    main()

File: scripts/gen-android-icons.js
// Regenerates the Android APK launcher icons — zero dependencies.
//
// Adapted icon (the one modern launchers draw): a transparent foreground layer
// holding the BARE mark, on the theme background color. The launcher masks the
// 108dp layer to whatever shape the device uses, so the foreground has to keep
// the artwork inside the 72dp safe circle — `npx tauri icon` derives it from
// the full-bleed badge instead, which the mask then clips.
//
// Legacy + round icons are the badge itself (public/icons/campfire-badge.png:
// the mark already on the theme-colored circle), so an old launcher and a
// Pixel's themed icon agree.
//
// Re-run after every `npx tauri icon ...` (see app/README.md "Icons").
const fs = require('fs');
const path = require('path');
const { decodePNG, encodePNG, resample } = require('./png-util');
const { compositeSquare } = require('./mark-render');

const root = path.join(__dirname, '..');
const markFile = path.join(root, 'public', 'icons', 'campfire-logo.png');
const badgeFile = path.join(root, 'public', 'icons', 'campfire-badge.png');
const { w: lw, h: lh, px: mark } = decodePNG(markFile);
const { w: bw, h: bh, px: badge } = decodePNG(badgeFile);
if (lw !== lh) throw new Error(`campfire-logo.png must be square, got ${lw}x${lh}`);
if (bw !== bh) throw new Error(`campfire-badge.png must be square, got ${bw}x${bh}`);

const THEME = [0x1a, 0x1d, 0x29]; // --bg / ic_launcher_background
const res = path.join(root, 'app', 'src-tauri', 'gen', 'android', 'app', 'src', 'main', 'res');

// Center an RGBA layer over transparency.
function placeTransparent(layer, box, size) {
  const out = Buffer.alloc(size * size * 4); // all zeros = transparent
  const off = Math.round((size - box) / 2);
  for (let y = 0; y < box; y++) {
    for (let x = 0; x < box; x++) {
      const s = (y * box + x) * 4;
      if (layer[s + 3] === 0) continue;
      const o = ((y + off) * size + (x + off)) * 4;
      out[o] = layer[s]; out[o + 1] = layer[s + 1]; out[o + 2] = layer[s + 2];
      out[o + 3] = layer[s + 3];
    }
  }
  return out;
}

const FOREGROUND = { mdpi: 108, hdpi: 162, xhdpi: 216, xxhdpi: 324, xxxhdpi: 432 };
const LEGACY = { mdpi: 48, hdpi: 72, xhdpi: 96, xxhdpi: 144, xxxhdpi: 192 };

for (const [dpi, size] of Object.entries(FOREGROUND)) {
  // 58% of the 108dp layer ≈ a 63dp mark inside the 72dp safe circle.
  const box = Math.round(size * 0.58);
  const layer = resample(mark, lw, lh, box, box);
  fs.writeFileSync(
    path.join(res, `mipmap-${dpi}`, 'ic_launcher_foreground.png'),
    encodePNG(size, size, placeTransparent(layer, box, size))
  );
  console.log('foreground', dpi, `${box}/${size}`);
}
for (const [dpi, size] of Object.entries(LEGACY)) {
  // The badge at full size. The plain legacy icon is the badge flattened onto
  // the theme color (older launchers show the square as-is), the round one
  // keeps the badge's transparent corners for launchers that mask it.
  const layer = resample(badge, bw, bh, size, size);
  fs.writeFileSync(
    path.join(res, `mipmap-${dpi}`, 'ic_launcher.png'),
    encodePNG(size, size, compositeSquare(THEME, layer, size, size, size))
  );
  fs.writeFileSync(path.join(res, `mipmap-${dpi}`, 'ic_launcher_round.png'), encodePNG(size, size, layer));
  console.log('legacy', dpi, `${size}/${size} round ${size}/${size}`);
}
console.log('android launcher icons regenerated ->', res);
